### PR TITLE
supports_distributed_data

### PR DESCRIPTION
**Pull request by Rombur**
Replace the static variable ```supports_distributed_data``` by the member function ```supports_distributed_data()```. This is an incompatible change but necessary because of the new vector interface where every vector derives from ```ReadWriteVector``` and/or ```VectorSpaceVector```. Given that until not long ago neither @drwells, @davydden, or me knew about this variable (#2580), I really doubt that we will break anyone's code. I also changed the documentation because the new vectors don't implement ```compress()```.

### CODE REVIEW

File: include/deal.II/lac/block_vector_base.h [hunk 1]
```diff
@@ -975,6 +964,15 @@ class BlockVectorBase : public Subscriptor
   void update_ghost_values () const;
 
   /**
+   * A variable that indicates whether this vector supports distributed data
+   * storage.
+   *
+   * For the current class, the variable equals the value declared for the
+   * type of the individual blocks.
+   */
+  bool supports_distributed_data() const;
```
Comment (tamiko) [inline include/deal.II/lac/block_vector_base.h]: Can we make ```supports_distributed_data``` a static member?

File: include/deal.II/lac/block_vector_base.h [hunk 2]
```diff
@@ -2264,6 +2262,17 @@ void BlockVectorBase<VectorType>::extract_subvector_to (ForwardIterator
...
 
+
+template <typename VectorType>
+inline
+bool BlockVectorBase<VectorType>::supports_distributed_data() const
+{
+  // Create a dummy vector of type VectorType to check
+  // supports_distributed_data even if BlockVectorBase contains no block
+  VectorType dummy;
+  return dummy.supports_distributed_data();
```
Comment (tamiko) [inline include/deal.II/lac/block_vector_base.h]: This would avoid the ```dummy``` here.

Comment (tamiko) [inline include/deal.II/lac/block_vector_base.h]: I can, for example, imagine situations where a default constructor for ```VectorType``` might not be available.

**Comment by tamiko** [comment]
Rest looks good.

File: include/deal.II/lac/block_vector_base.h (hunk 1 above)
Comment (Rombur) [inline include/deal.II/lac/block_vector_base.h]: The problem is that if we do that ```VectorSpaceVector``` cannot have this variable define because the variable will be true for some derived classes and false for others. ```VectorSpaceVector``` is an abstract class so I guess it could work but we would need to downcast every time we would need to know the variable.

File: include/deal.II/lac/block_vector_base.h (hunk 2 above)
Comment (Rombur) [inline include/deal.II/lac/block_vector_base.h]: I know it's not ideal but I thought that it would be pretty strange to have a vector that doesn't have a default constructor.

**Comment by drwells** [comment]
This should be fine. I agree that it is unlikely that there are usages of this outside of generic programming inside the library itself, so I think we can be bad and remove it without a deprecation warning.

**Comment by drwells** [comment]
@bangerth According to the repository history you were the original author of that static variable (this was long before the new linear algebra base classes). Does this change look good to you?

File: doc/news/changes.h [hunk 3]
```diff
@@ -39,6 +39,12 @@ inconvenience this causes.
 
 <ol>
 
+ <li> changed: The static variable supports_distributed_data has been replaced
```
Comment (bangerth) [inline doc/news/changes.h]: Capitalize 'changed'

Comment (bangerth) [inline doc/news/changes.h]: I think you ought to also say that this variable is part of the vector interface, same for the function name in the next line.

File: include/deal.II/lac/block_vector_base.h (hunk 2 above)
Comment (bangerth) [inline include/deal.II/lac/block_vector_base.h]: This makes no sense to me. The way you wrote this function, it doesn't depend on the state of the block vector -- the function may as well be `static`. I think what you probably want is to loop over all blocks and combine what *they* report for this function in some useful way. If there are no blocks, you might as well throw an exception.

**Comment by bangerth** [comment]
I've got no objections to making this a function, though I think that the current implementation in `BlockVector` is wrong.

File: include/deal.II/lac/block_vector_base.h (hunk 2 above)
Comment (Rombur) [inline include/deal.II/lac/block_vector_base.h]: Right now the variable is ```static``` and totally independent of the state of vector. I don't see how changing the ```static``` variable to a ```static``` function will help with the inheritance problem in the new ```Vector``` interface. The behavior of the function is exactly the same as the behavior of the variable.

File: include/deal.II/lac/block_vector_base.h (hunk 1 above)
Comment (tamiko) [inline include/deal.II/lac/block_vector_base.h]: @Rombur Never mind. I now see what you try to achieve and how you want to use ```supports_distributed_data``` - I agree that a static member function will be awkward to use in case of inheritance and working with base classes.

File: include/deal.II/lac/block_vector_base.h (hunk 2 above)
Comment (tamiko) [inline include/deal.II/lac/block_vector_base.h]: @Rombur I agree with @bangerth here. If ```VectorType``` is an entity with a dynamic ```supports_distributed_data``` property, you have to check every block on runtime.

If this property, on the other hand is type bound, I would suggest, to either use a static member function, or - better - a type trait:
```
SupportsDistributedData<VectorType>::value
```

Comment (Rombur) [inline include/deal.II/lac/block_vector_base.h]: @tamiko the property is type bound, it only depends on the template ```VectorType```. That's why, right now, it is a static variable...
 - I don't see how a static member function will work. Unless we decide that the function is not part of  ```VectorSpaceVector``` which will necessitate to cast down the derived ```Vector``` all the time. Moreover, if in the future someone wants to create a distributed vector which derives from ```ReadWriteVector```, it won't work. 
 - isn't type trait C++11?

Comment (tamiko) [inline include/deal.II/lac/block_vector_base.h]: OK, I do not want to stall this PR. My fundamental issue - or, more precisely, the thing I don't fully understand is the following:

- You change ```supports_distributed_data``` to a virtual member function ```supports_distributed_data``` to allow for (runtime) type polymorphism. I.e., the following is possible:
```
VectorBase *pointer = new VectorSpecialized;
pointer->supports_distributed_data();
```
I mean, this is in line with the abstract interfaces you defined for our linear algebra.

 - But, on the other hand you use it exclusively "type bound", i.e. your implementation above has the implicit assumption that creation of above ```VectorType``` is possible _and_ is a meaningful type. But what about a ```BlockVectorBase<VectorSpaceVector>```? [1]

A lot of this confusion is due to the fact that we happen to introduce a second form of polymorphism to an interface that formerly only had one. We now have:
 - Runtime type polymorphism due to inheritance from interface classes
 - Static type polymorphism due to templates

[1] The point is, we're already half way through transforming our linear algebra classes (at least for vectors) mainly to the first type of polymorphism---even though we use it mainly as a workaround for the lack of "concepts" in C++11/14/17. But by doing that we should support it completely, i.e., ```BlockVectorBase``` should ask its concrete block objects on runtime.

Comment (tamiko) [inline include/deal.II/lac/block_vector_base.h]: Or, alternatively, if ```VectorSpaceVector``` (and similar) are meant to function as some sort of C++ concept to enforce a uniform interface - and we happen to not use the runtime polymorphism at all - then, we should use it strictly only on the template level. For example:
```
template<typename Vector2,
         typename = typename std::enable_if<std::is_base_of<VectorBaseVector<Number> Vector2<Number>>::value>::type>
Vector<number> &operator+=(const Vector2<Number> &V);
```
(This looks horrible, I know...)

Comment (Rombur) [inline include/deal.II/lac/block_vector_base.h]: This is pretty much how it works right now. If you try to add two vectors which are derived from ```VectorSpaceVector``` but belong to different derived classes, you will get a runtime error. Nothing forbids to implement a case where you could add two vectors of different types but there is no plan to do so.

Comment (bangerth) [inline include/deal.II/lac/block_vector_base.h]: @tamiko -- `BlockVector` actually knows the exact type of the underlying vector types for the individual blocks, because it allocates them. They're equal to the template type. In other words, we know for sure that when @Rombur uses
```
  VectorType dummy;
  return dummy.supports_distributed_data();
```
that dummy has the right type.

But there is still the original point I don't understand and that I haven't understood from @Rombur 's comments. If you argue that this now needs to be a function rather than a static variable, then this only makes sense if one class may either return `true` or `false`, depending on what its current internal state is. Otherwise, if it does not depend on the internal state (i.e., the values of member variables), then it could as well be a static const variable. But if, indeed, the returned value may depend on the state of the variables, then necessarily the `BlockVector` class cannot just create an empty object, but actually needs to query the individual blocks. Isn't this right?

(The other interpretation is that you want to access the variable through the *base class* but you don't know what derived class you have. In that case, the function will be abstract virtual. I may be missing this, but I don't actually see a place where you use this right now. Do you have a use case in mind where you only get a pointer-to-base, but still need to know whether the vector supports distributed data?)

Comment (tamiko) [inline include/deal.II/lac/block_vector_base.h]: @bangerth For example have a look at ```include/deal.II/la_parallel_vector.templates.h``` quite a number of ```operator```s have a signature like this:
```
925     template <typename Number>                                                   
926     Vector<Number> &                                                             
927     Vector<Number>::operator += (const VectorSpaceVector<Number> &vv)            
```

**Comment by kronbichler** [comment]
I'm late to this discussion, but let me add my point of view:

My main issue with this is that I do not see why we need such a function in the general vector interfaces. We currently use it in only two places inside the library (the function in `numerics/vector_tools.templates.h` that was recently added in 2580 and once in `lac/constraint_matrix.templates.h` for `distribute(VectorType)` where we try to be smart and avoid some copy operation. Those two places do not convince me that it is worth to address the ambiguity problem that we definitely have at this point in terms of runtime polymorphism vs. static polymorphism.

In other words, my issue is that the vector interfaces should uniquely determine on what kinds of operations we support for what kinds of vectors (readwrite vs VectorSpaceVector). Unless there are performance reasons that conflict with this, which I do not see here.

File: include/deal.II/lac/block_vector_base.h (hunk 2 above)
Comment (Rombur) [inline include/deal.II/lac/block_vector_base.h]: @bangerth like @tamiko said we use the base class quite a lot in the interface that's why I want to have it there.

Comment (bangerth) [inline include/deal.II/lac/block_vector_base.h]: @tamiko  I just wanted to point out that in the concrete implementation of `BlockVectorBase<VectorType>::supports_distributed_data()`, we actually do know the concrete type of the objects.

@Rombur I see. Makes sense.

**Comment by Rombur** [comment]
I have a branch where I instantiate ```LinearAlgebra::Vector``` and the last problem to fix is this ```supports_distributed_data``` variable. Do we agree this is the right way to change ```supports_distributed_data``` or should we get ride of the variable like @kronbichler suggested?

**Comment by bangerth** [comment]
I'm for the latter, i.e., getting rid of the variable. If that is awkward for the moment, then I still think it should be a static property of the class, i.e., a static const member variable; if a class doesn't know up front whether or not it will have distributed data then it should err on the safe side.
